File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* kthnode(ListNode* head,int k){
        ListNode* temp=head;
        while(temp!=nullptr && k>1){
            temp=temp->next;
            k--;
        }
        return temp;
    }

    ListNode* reverse(ListNode* head){
        ListNode* prev=nullptr,*curr=head,*next=nullptr;
        while(curr){
            next=curr->next;
            curr->next=prev;
            prev=curr;
            curr=next;
        }
        return prev;
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* prevnode=nullptr,*nextnode=nullptr;
        ListNode*  temp=head;
        while(temp){
            ListNode* kth=kthnode(temp,k);
            if(kth==nullptr){
                if(prevnode==nullptr) return head;
                prevnode->next=temp;
                return head;
            }
            nextnode=kth->next;
            kth->next=nullptr;
            ListNode* newhead=reverse(temp);
            if(temp==head){
                head=kth;
            }
            
            if(prevnode) prevnode->next=newhead;

            prevnode=temp;
            temp=nextnode;
        }
        return head;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {

    public ListNode kthNode(ListNode head,int k){
        while(head!=null && k>0){
            head=head.next;
            k--;
        }
        return head;
    }

    public ListNode reverse(ListNode head){
        ListNode prev=null,curr=head;
        while(curr!=null){
            ListNode next=curr.next;
            curr.next=prev;
            prev=curr;
            curr=next;
        }
        return prev;
    }

    public ListNode ReverseKGroup(ListNode head, int k) {
        ListNode temp=head;
        ListNode prevNode=null,nextNode=null;

        while(temp!=null){
            ListNode knode=kthNode(temp,k-1);
            if(knode==null){
                if(prevNode!=null) prevNode.next=temp;
                break;
            }
            nextNode=knode.next;
            knode.next=null;
            ListNode rev=reverse(temp);
            if(temp==head) head=knode;
            if(prevNode!=null) prevNode.next=rev;
            prevNode=temp;
            temp=nextNode;
        }
        return head;
    }
}